src/az_ulib_ustream_forward/az_ulib_ustream_forward.c

```diff
@@ -0,0 +1,171 @@
+    az_ulib_flush_callback flush_callback, 
+    az_ulib_callback_context flush_callback_context)
+{
+  // precondition checks
+  _az_PRECONDITION_NOT_NULL(ustream_forward);
+  _az_PRECONDITION(AZ_ULIB_USTREAM_FORWARD_IS_TYPE_OF(ustream_forward, api));
+  _az_PRECONDITION_NOT_NULL(flush_callback);
+  _az_PRECONDITION_NOT_NULL(flush_callback_context);
+
+  AZ_ULIB_TRY
+  {
+    // point to data from instance
+    size_t buffer_size;
+    AZ_ULIB_THROW_IF_AZ_ERROR(concrete_get_size(ustream_forward, &buffer_size));
+    IGNORE_CAST_QUALIFICATION
+    const uint8_t* buffer = (uint8_t*)ustream_forward->_internal.ptr;
+    RESUME_WARNINGS
+
+    // invoke callback
+    (*flush_callback)(buffer, buffer_size, flush_callback_context);
+  }
+  AZ_ULIB_CATCH (...) {}
+
+  return AZ_ULIB_TRY_RESULT;
+}
+
+static az_result concrete_read(
+    az_ulib_ustream_forward* ustream_forward,
+    uint8_t* const buffer,
+    size_t buffer_length,
+    size_t* const size)
+{
+  _az_PRECONDITION(AZ_ULIB_USTREAM_FORWARD_IS_TYPE_OF(ustream_forward, api));
+  _az_PRECONDITION_NOT_NULL(buffer);
+  _az_PRECONDITION(buffer_length > 0);
+  _az_PRECONDITION_NOT_NULL(size);
+
+  az_result result;
+
+  if (ustream_forward->_internal.inner_current_position >= ustream_forward->_internal.length)
+  {
+    *size = 0;
+    result = AZ_ULIB_EOF;
+  }
+  else
+  {
+    size_t remain_size
+        = ustream_forward->_internal.length - (size_t)ustream_forward->_internal.inner_current_position;
+    *size = (buffer_length < remain_size) ? buffer_length : remain_size;
+    IGNORE_MEMCPY_TO_NULL
+    memcpy(
+        buffer,
+        (const uint8_t*)ustream_forward->_internal.ptr + ustream_forward->_internal.inner_current_position,
+        *size);
+    RESUME_WARNINGS
+    ustream_forward->_internal.inner_current_position += *size;
+    result = AZ_OK;
+  }
+
+  return result;
+}
+
+static az_result concrete_get_size(az_ulib_ustream_forward* ustream_forward, size_t* const size)
+{
+  _az_PRECONDITION(AZ_ULIB_USTREAM_FORWARD_IS_TYPE_OF(ustream_forward, api));
+  _az_PRECONDITION_NOT_NULL(size);
+
+  *size = ustream_forward->_internal.length - ustream_forward->_internal.inner_current_position;
+
+  return AZ_OK;
+}
+
+static az_result concrete_dispose(az_ulib_ustream_forward* ustream_forward)
+{
+  _az_PRECONDITION(AZ_ULIB_USTREAM_FORWARD_IS_TYPE_OF(ustream_forward, api));
+
+  if (ustream_forward->_internal.data_release)
+  {
+    /* If `data_relese` was provided is because `ptr` is not `const`. So, we have an Warning
+     * exception here to remove the `const` qualification of the `ptr`. */
+    IGNORE_POINTER_TYPE_QUALIFICATION
+    ustream_forward->_internal.data_release(ustream_forward->_internal.ptr);
+    RESUME_WARNINGS
+  }
+  if (ustream_forward->_internal.ustream_forward_release)
+  {
+    ustream_forward->_internal.ustream_forward_release(ustream_forward);
+  }
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_ulib_ustream_forward_init(
```

Comment:
I think it do not make sense to have 2 .h implemented by a single .c, so how about unify `az_ulib_ustream_forward.h` and `az_ulib_ustream_forward_base.h` in a single `az_ulib_ustream_forward.h`?

Reply:
I thought the same thing. I was following the layout of ustream, but this makes much more sense.